cmd/admin-handlers.go

```diff
@@ -727,41 +824,139 @@ func (a adminAPIHandlers) SetConfigHandler(w http.ResponseWriter, r *http.Reques
+	peersErrs, err := a.setConfigDotJSON(ctx, configBytes)
+	if err != nil {
+		writeErrorResponseJSON(w, toAPIErrorCode(err), r.URL)
+		return
+	}
 
-	// Check if the operation succeeded in quorum or more nodes.
-	rErr := reduceWriteQuorumErrs(ctx, errs, nil, len(globalAdminPeers)/2+1)
-	if rErr != nil {
-		writeSetConfigResponse(w, globalAdminPeers, errs, false, r.URL)
+	if reduceWriteQuorumErrs(ctx, peersErrs, nil, len(globalAdminPeers)/2+1) != nil {
+		writeSetConfigResponse(w, globalAdminPeers, peersErrs, false, r.URL)
 		return
 	}
 
-	// Take a lock on minio/config.json. NB minio is a reserved
-	// bucket name and wouldn't conflict with normal object
-	// operations.
-	configLock := globalNSMutex.NewNSLock(minioReservedBucket, minioConfigFile)
-	if configLock.GetLock(globalObjectTimeout) != nil {
-		writeErrorResponseJSON(w, ErrOperationTimedOut, r.URL)
+	// serverMux (cmd/server-mux.go) implements graceful shutdown,
+	// where all listeners are closed and process restart/shutdown
+	// happens after 5s or completion of all ongoing http
+	// requests, whichever is earlier.
+	writeSetConfigResponse(w, globalAdminPeers, peersErrs, true, r.URL)
+
+	// Restart all node for the modified config to take effect.
+	sendServiceCmd(globalAdminPeers, serviceRestart)
+}
+
+func convertType(str string, jsonType gjson.Type) (interface{}, error) {
+	switch jsonType {
+	case gjson.False, gjson.True:
+		return strconv.ParseBool(str)
+	case gjson.JSON:
+		return gjson.Parse(str).Value(), nil
+	case gjson.String:
+		return str, nil
+	case gjson.Number:
+		return strconv.ParseFloat(str, 64)
+	default:
+		return nil, nil
+	}
+}
+
+// SetConfigKeysHandler - PUT /minio/admin/v1/config-keys
+func (a adminAPIHandlers) SetConfigKeysHandler(w http.ResponseWriter, r *http.Request) {
+
+	ctx := context.Background()
+	// Get current object layer instance.
+	objectAPI := newObjectLayerFn()
+	if objectAPI == nil {
+		writeErrorResponseJSON(w, ErrServerNotInitialized, r.URL)
 		return
 	}
-	defer configLock.Unlock()
 
-	// Rename the temporary config file to config.json
-	errs = commitConfigPeers(globalAdminPeers, tmpFileName)
-	rErr = reduceWriteQuorumErrs(ctx, errs, nil, len(globalAdminPeers)/2+1)
-	if rErr != nil {
-		writeSetConfigResponse(w, globalAdminPeers, errs, false, r.URL)
+	// Deny if WORM is enabled
+	if globalWORMEnabled {
+		writeErrorResponseJSON(w, ErrMethodNotAllowed, r.URL)
+		return
+	}
+
+	// Validate request signature.
+	adminAPIErr := checkAdminRequestAuthType(r, "")
+	if adminAPIErr != ErrNone {
+		writeErrorResponseJSON(w, adminAPIErr, r.URL)
+		return
+	}
+
+	configBytes, err := a.getConfigDotJSON()
+	if err != nil {
+		writeErrorResponseJSON(w, toAPIErrorCode(err), r.URL)
+		return
+	}
+
+	origConfig := string(configBytes)
+	configStr := origConfig
+
+	queries := r.URL.Query()
+
+	for k, v := range queries {
+		var elem string
+		if len(v) > 0 {
+			elem = v[0]
+		}
+
+		jsonFieldType := gjson.Get(origConfig, k).Type
```

Comment:
In current implementation, we try to find the user-provided key name in the current config.json file and assume user-provided key/value pair to be valid. If the key name does not exist in the config.json file, it is still assumed to be a valid new key/value entry, but the value is ignored and saved as `null`, since `gjson.Get(origConfig, k).Type` function returns `null` for the type and hence the value is determined and returned  as `null` from `convertType` function.

IMHO, we need to change this logic.

First of all, we should not depend on config.json file, which not necessarily has all possible valid key names, to check and validate key names and their value types.  Instead, we should validate key names against config data type struct and check if the user-provided key name is legit. If the key name cannot be found in the config type struct, then mc command should fail and error out with `"Invalid key"` error message. If valid, then we can pull its value type from config data type struct and feed it into `convertType` function. If `convertType` function falls into the `switch default` section, that is if function cannot return a valid value, this should mean the user-provided value type is wrong and mc command should fail and error out with `Invalid value type` error message. 
If all good, the user-provided key/value pair can be added into the config.json file either by modifying the value of the existing key if the key is already part of the config.json or by appending it as a new entry otherwise.